### ISSUE TITLE
`byteArrayFromString` produces incorrect result for short string

### ISSUE THREAD

**Issue by DelevoXDG** [starknet-io/starknet.js#1003]
**Describe the bug**

The algorithm, used both in [byteArray.ts](https://github.com/starknet-io/starknet.js/blob/fdcc7fbf5957c14df765fd78508db6539b6124e3/src/utils/calldata/byteArray.ts#L48) and [typedData.ts](https://github.com/starknet-io/starknet.js/blob/6cd35aa26d47d4f3a594349d65792fd1f368ba67/src/utils/typedData.ts#L65) seems to produce incorrect results for short strings of length <31. 

It ensures at least one value (`0x00`) is present in `ByteArray`'s `data`, contrary to the [docs](https://docs.starknet.io/documentation/architecture_and_concepts/Smart_Contracts/serialization_of_Cairo_types/#serialization_of_byte_arrays), where in such case `data` should be empty.

**To Reproduce**
Run the following test, based on the example from the docs:
```js
  test('convert string to ByteArray', () => {
    expect(byteArray.byteArrayFromString('hello')).toEqual({
      data: [],
      pending_word: '0x68656c6c6f',
      pending_word_len: 5,
    });
  });
```
**Expected behavior**
Test passes

**Actual behavior**
Test fails, actual `ByteArray` is:
```
data: ['0x00'],
pending_word: '0x68656c6c6f',
pending_word_len: 5,
```

**Desktop (please complete the following information):**
- Starknet.js version fdcc7fbf5957c14df765fd78508db6539b6124e3

**Comment by tabaktoni** [comment]
Thank you for reporting, we will look into it. @PhilippeR26

**Comment by PhilippeR26** [comment]
Hello,
This documentation is not conform to the real behavior of Starknet.
When Starknet serializes a `ByteArray` including less than 31 characters, the result is 
```
[
  '0x1',
  '0x0',
  '0x54616b6520636172652e205a6f7267206973206261636b',
  '0x17'
]
```
 `'0x1','0x0',` means an array of one element (value 0).
 
 Starknet.js will continue to follow the real behavior of Starknet network.

**Comment by DelevoXDG** [comment]
Thanks for your reply. 

Can you please clarify how did you confirm that?

**Comment by PhilippeR26** [comment]
I just deployed in Sepolia the Starknet.js test contract  at : `0x660edd51a76b970cc8b843e5e0560e48afa2bb73e2530ab9a7425b17c1366b2`
If you launch this :
```typescript
const resp4= await stringContract.call("proceed_string",["Take care."],{parseResponse:false});
```
The response is :
```
0x1,0x0,0x54616b6520636172652e205a6f7267206973206261636b,0x17
```

**Comment by amanusk** [comment]
The issue seems to be with the way data is serialized and sent when calling `proceed_string`.

Calling the following function in Cairo/snforge returns the data array as explained in the doc
```        
fn proceed_array(self: @ContractState, mess: ByteArray) -> Array<felt252> {
            let mut res = mess;
            let add: ByteArray = " Zorg is back";
            res.append(@add);

            let mut serialized_string: Array<felt252> = Default::default();
            res.serialize(ref serialized_string);
            serialized_string
        }
```
`println` will return [0, 1614488889802844162655565249641883848014914411, 19]

But, calling the function in the test, from sn.js returns an array with "0x0" as a first element.

To confirm this, a function w/o any appending will return the correct structure.
```
        fn default_string(self: @ContractState) -> ByteArray {
            let res: ByteArray = "Zorg is back";
            res
        }

```

calling it from sn.js, results in the following result
```
  let resp4 = await myTestContract.call("default_string", [], { parseResponse: false });
  console.log("Default sting");
  console.log(resp4);
```

will result in `[ '0x0', '0x5a6f7267206973206261636b', '0xc' ]`, as expected and described in docs.

TL;DR, ByteArray should be serialized as explained in the docs, but sn.js is serializing them incorrectly when sending them to a contract
